tests/test_warnings.cpp

```diff
@@ -0,0 +1,77 @@
+/*
+    tests/test_warnings.cpp -- usage of raise_warning() and warnings categories
+
+    Copyright (c) 2023 Jan Iwaszkiewicz <jiwaszkiewicz6@gmail.com>
+
+    All rights reserved. Use of this source code is governed by a
+    BSD-style license that can be found in the LICENSE file.
+*/
+
+#include "pybind11_tests.h"
+
+#include <utility>
+
+namespace warning_helpers {
+void warn_function(py::module &m, const char *name, py::handle category, const char *message) {
+    m.def(name, [category, message]() { py::raise_warning(message, category); });
+}
+} // namespace warning_helpers
+
+class CustomWarning {};
+
+TEST_SUBMODULE(warnings_, m) {
+
+    // Test warning mechanism base
+    m.def("raise_and_return", []() {
+        std::string message = "Warning was raised!";
+        py::raise_warning(message.c_str(), py::warnings::warning_base);
+        return 21;
+    });
+
+    m.def("raise_default", []() { py::raise_warning("RuntimeWarning is raised!"); });
+
+    m.def("raise_from_cpython",
+          []() { py::raise_warning("UnicodeWarning is raised!", PyExc_UnicodeWarning); });
+
+    m.def("raise_and_fail",
+          []() { py::raise_warning("RuntimeError should be raised!", PyExc_Exception); });
+
+    // Test custom warnings
+    static py::warning<CustomWarning> my_warning(m, "CustomWarning", py::warnings::deprecation);
```

Comment:
This is a very common pitfall:

The destructor may run (and call `Py_DECREF`) only after the Python interpreter is finalized already.

Best to adopt this approach:

https://github.com/pybind/pybind11/blob/43de8014f96ab8fc0028f11a26edbcf9b937cd91/tests/test_exceptions.cpp#L119-L121